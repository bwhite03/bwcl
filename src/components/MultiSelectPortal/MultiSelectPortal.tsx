import React from 'react';
import { createPortal } from 'react-dom';

export interface MultiSelectProps<T> {
  data: T[];
  keyDescriptor: string;
  optionValue: keyof T;
  optionText: keyof T;
  open: boolean;
  handleCheck: (checkbox: React.ChangeEvent<HTMLInputElement>, data: T) => void;
  handleSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  search: string;
  amIChecked: (e: T) => boolean;
  inputId: string;
  id: string;
}

function MultiSelectPortal<T>(props: MultiSelectProps<T>) {
  const {
    open,
    handleSearchChange,
    search,
    inputId,
    data,
    keyDescriptor,
    optionValue,
    optionText,
    amIChecked,
    handleCheck,
    id = 'multi-select-checkboxes',
  } = props;
  const portalDiv = document.getElementById('ms-portal-container');

  return (
    <React.Fragment>
      {open && portalDiv
        ? createPortal(
            <div className="multi-select-checkboxes" id={id}>
              <input
                className="form-control"
                placeholder="search"
                type="text"
                onChange={handleSearchChange}
                value={search}
                id={inputId}
              />
              {data && data.length > 0 ? (
                <React.Fragment>
                  {data.map((d, i) => (
                    <label key={`${keyDescriptor}-${i}`}>
                      <input
                        type="checkbox"
                        checked={amIChecked(d)}
                        onChange={(e) => handleCheck(e, d)}
                      />
                      {d[optionText]}
                    </label>
                  ))}
                </React.Fragment>
              ) : null}
            </div>,
            portalDiv
          )
        : null}
    </React.Fragment>
  );
}

export default MultiSelectPortal;
