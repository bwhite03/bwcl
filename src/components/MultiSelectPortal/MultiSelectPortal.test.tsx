import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import MultiSelectPortal from './MultiSelectPortal';

const data = [
  {
    id: 1,
    value: 'Grocery',
  },
  {
    id: 2,
    value: 'Nonfood',
  },
  {
    id: 3,
    value: 'HBA',
  },
  {
    id: 4,
    value: 'Meat',
  },
  {
    id: 5,
    value: 'Meat by #',
  },
  {
    id: 6,
    value: 'Produce',
  },
  {
    id: 7,
    value: 'Produce by #',
  },
];

describe('MultiSelectPortal', () => {
  test('should render correctly', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);
    const portal = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(portal);

    const handleCheck = jest.fn();
    const handleSearchChange = jest.fn();
    const amIChecked = jest.fn();

    const { container } = render(
      <MultiSelectPortal
        data={data}
        keyDescriptor={'select a department'}
        //@ts-ignore
        optionValue="id"
        //@ts-ignore
        optionText="value"
        open={true}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        amIChecked={amIChecked}
        search=""
        inputId="inputid"
        id="id"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should render correctly without an id', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);

    const handleCheck = jest.fn();
    const handleSearchChange = jest.fn();
    const amIChecked = jest.fn();

    const { container } = render(
      // @ts-ignore
      <MultiSelectPortal
        data={data}
        keyDescriptor={'select a department'}
        //@ts-ignore
        optionValue="id"
        //@ts-ignore
        optionText="value"
        open={true}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        amIChecked={amIChecked}
        search=""
        inputId="inputid"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should not render a portal', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);

    const handleCheck = jest.fn();
    const handleSearchChange = jest.fn();
    const amIChecked = jest.fn();

    const { container } = render(
      // @ts-ignore
      <MultiSelectPortal
        data={data}
        keyDescriptor={'select a department'}
        //@ts-ignore
        optionValue="id"
        //@ts-ignore
        optionText="value"
        open={false}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        amIChecked={amIChecked}
        search=""
        inputId="inputid"
        id="id"
      />
    );

    expect(container).toMatchSnapshot();
  });

  test('should render with no data', () => {
    // const div = document.createElement("div");
    // div.id = "ms-portal-container";
    // document.body.appendChild(div);

    const handleCheck = jest.fn();
    const handleSearchChange = jest.fn();
    const amIChecked = jest.fn();

    const { container } = render(
      // @ts-ignore
      <MultiSelectPortal
        data={[]}
        keyDescriptor={'select a department'}
        //@ts-ignore
        optionValue="id"
        //@ts-ignore
        optionText="value"
        open={true}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        amIChecked={amIChecked}
        search=""
        inputId="inputid"
        id="id"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should check a box', () => {
    const div = document.createElement('div');
    div.id = 'ms-portal-container';
    document.body.appendChild(div);

    const handleCheck = jest.fn();
    const handleSearchChange = jest.fn();
    const amIChecked = jest.fn();

    const { container, getByRole } = render(
      // @ts-ignore
      <MultiSelectPortal
        data={[]}
        keyDescriptor={'select a department'}
        //@ts-ignore
        optionValue="id"
        //@ts-ignore
        optionText="value"
        open={true}
        handleCheck={handleCheck}
        handleSearchChange={handleSearchChange}
        amIChecked={amIChecked}
        search=""
        inputId="inputid"
        id="id"
      />
    );

    // const check = getByRole("checkbox", { name: /Grocery/ });
    // fireEvent.click(check);
  });
});
