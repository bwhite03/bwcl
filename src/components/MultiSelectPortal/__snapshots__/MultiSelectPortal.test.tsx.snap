// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectPortal should not render a portal 1`] = `<div />`;

exports[`MultiSelectPortal should render correctly 1`] = `<div />`;

exports[`MultiSelectPortal should render correctly without an id 1`] = `<div />`;

exports[`MultiSelectPortal should render with no data 1`] = `<div />`;
